src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb

```diff
@@ -2594,12 +2594,60 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
         End Function
 
         Protected Overrides Function CommonCreateTupleTypeSymbol(elementTypes As ImmutableArray(Of ITypeSymbol), elementNames As ImmutableArray(Of String)) As INamedTypeSymbol
-            Dim vbElementTypes = elementTypes.SelectAsArray(Function(t) t.EnsureVbSymbolOrNothing(Of TypeSymbol)(NameOf(elementTypes)))
+            If elementTypes.IsDefault Then
+                Throw New ArgumentNullException(NameOf(elementTypes))
+            End If
+
+            If elementTypes.Length <= 1 Then
+                Throw New ArgumentException(CodeAnalysisResources.TuplesNeedAtLeastTwoElements, NameOf(elementNames))
+            End If
+
+            CheckTupleElementNames(elementTypes.Length, elementNames)
+
+            Dim typesBuilder = ArrayBuilder(Of TypeSymbol).GetInstance(elementTypes.Length)
+            For i As Integer = 0 To elementTypes.Length - 1
+                If elementTypes(i) Is Nothing Then
+                    Throw New ArgumentNullException($"{NameOf(elementTypes)}[{i}]")
+                End If
 
-            Return TupleTypeSymbol.Create(locationOpt:=Nothing, elementTypes:=vbElementTypes, elementLocations:=Nothing, elementNames:=elementNames, compilation:=Me)
+                typesBuilder.Add(elementTypes(i).EnsureVbSymbolOrNothing(Of TypeSymbol)($"{NameOf(elementTypes)}[{i}]"))
+            Next
```

Comment:
The argument checking is split across two methods and we're constructing `paramName` for each element, even in the success case. Consider creating a helper method that checks and converts the entire array: `Dim types = EnsureVbSymbols(Of TypeSymbol)(elementNames, NameOf(elementNames))`. I believe the same method was needed in another PR recently.